ENH: Implement Multivariate Rejection Sampling (MRS)

Lucas-Prates opened the pull request:
## Pull request type

- [X] Code changes (bugfix, features)

## Checklist
<!-- Remove irrelevant items to this PR. -->

- [x] Unit tests for the changes have been added
- [x] Integration tests for the changes have been added
- [x] Docs have been reviewed and added / updated
- [x] Lint (`black rocketpy/ tests/`) has passed locally 
- [x] All tests (`pytest tests -m slow --runslow`) have passed locally
- [x] `CHANGELOG.md` has been updated (if relevant)
- [x] RST documentation
- [x] Monte Carlo comparison Feature

## New behavior
This PR implements the MRS requested in #162 and described in [RocketPy paper](https://ascelibrary.org/doi/10.1061/%28ASCE%29AS.1943-5525.0001331).

## Breaking change

- [x] Yes (perhaps)

## Additional information

Lucas-Prates commented:
This PR is ready for a first "Design Review." I would like to get your opinion if this implementation provides what you think on how the user should use the MRS.

Albeit a implementation as a function seems natural, I implemented as a class because:
1) the function would be humongous;
2) if the user wants to resample several times, the data monte carlo data is only read once from the harddrive.

It currently works as follows:

0) Input: monte carlo filepath prefix, mrs filepath prefix, distribution dictionary;
1) Load input and output data from a monte carlo simulation into memory (python objects - lists of jsons);
2) To avoid having to read data twice, while loading, precompute some important properties required in the sampler algorithm;
3) Select and save iteratively accepted samples;
4) Output: files are saved in the same "scheme" as the `MonteCarlo` simulation.

I provided a quick and dirty notebook, which will be removed, just to show how the class is being used at the moment.

Gui-FernandesBR reviewed (commented):
I don't have much time so I'll be short

* The class dies after you sample the data once, this makes it pointless to have a class
* Instead I'd remove the sample dictionary from arguments.
* First read the data during initialization. Then you use a function to set the variables you are going to allow be varied. This way u can already anticipate which variables may be varied.
* We want almost instant results for a MonteCarlo simulation after using MRS.
* Other thing is that the user must supply the original pdf. Could we possile estimate this from data? (imagine 70k)
* Finally, plotting is crucial for MRS, or even tables. We'd love to see more on that later.

Lucas-Prates commented on rocketpy/simulation/multivariate_rejection_sampler.py:
```diff
@@ -0,0 +1,220 @@
+
+        # compute sup ratio
+        for line_input_json, line_output_json in zip(self.input_list, self.output_list):
+            acceptance_prob = 1 / self.sup_ratio  # probability the sample is accepted
```
The only way that `self.sup_ratio` could be zero is if **all** the density ratios are zero. In that case, the user probably specified two distributions with different supports. I guess it makes sense to validate that `self.sup_ratio` is not zero after it is computed.

phmbressan commented on rocketpy/simulation/multivariate_rejection_sampler.py:
```diff
@@ -0,0 +1,259 @@
+                f"the monte carlo input file {input_filename}!"
+            ) from e
+
+        input_file.close()
```
This line should be inside a `finally` block, I believe. The same applies to the method ``load_output``.

phmbressan commented on rocketpy/simulation/multivariate_rejection_sampler.py:
```diff
@@ -0,0 +1,259 @@
+        """Loads input information from monte carlo in a SampleInformation
+        object
+        """
+        input_filename = f"{self.monte_carlo_filepath}.inputs.txt"
```
If possible, it is better practice to handle paths with the `pathlib` standard library. Check out the parallel monte carlo PR as an example of handling the `filename` in the MonteCarlo class.

Lucas-Prates commented:
This PR is ready for review again. Changes from last time:

1. addressed some of the previous suggestions of @phmbressan;
2. added a `comparison_info` method that provides a statistical summary print comparison of the results of two monte carlo simulations, similar to the method used in the regular monte carlo class;
3. added a `comparison_plots` method that plots boxplots and histograms comparison of the results of two monte carlo simulations;
4. added a `compare_ellipses` method that plots the ellipses for the apogee and landing point comparison of the results of two monte carlo simulations;
5. added the `mrs.rst` file to the documentation explaining what is the MRS and how to use it;
6. added the `MultivariateRejectionSampler` class .rst documentation;
7. removed the `test_mrs.ipynb` notebook since the .rst file does the same in a much clearer way. 

For review, I recommend mostly checking the .rst documentation. Here are some previews of the comparisons if you do not have the time to compile the html:

Print comparison: 
![image](https://github.com/user-attachments/assets/bcc64f69-c78f-4893-9631-e829039f57a8)

Plot comparison:
![image](https://github.com/user-attachments/assets/aeaf6228-0d7f-4459-b724-ba8f337ec21f)

Ellipses comparison:
![image](https://github.com/user-attachments/assets/6e5424db-cf9e-43b2-bbef-0c8297aed969)

Lucas-Prates commented:
A second comment: I made the design choice to implement the comparison methods in the `MonteCarlo` class instead of the `MultivariateRejectionSampler`. Here are the most relevant arguments:

1. It makes sense to compare monte carlo simulations even outside of MRS, i.e. even one is not a sub-sample of the other. Hence, implementing this inside the MRS would make this usage extremely odd;
2. It keeps the `MultivariateRejectionSampler` class implementation really simple and with only one job: **to sample**. 

One option is, of course, to create another class for that job, but it did not seem to be the best option.

Gui-FernandesBR commented on docs/reference/classes/MultivariateRejectionSampler.rst:
A few more commentaries that you may or may not accept in order to improve readability:

```text
"but their distribution matches the one newly specified without having to re-run the the simulation." → "but its distribution matches the newly specified one without having to re-run the simulation."
(Fixes double "the" and improves sentence flow.)

"The time efficiency of the MRS is specially interesting in two scenarios" → "The time efficiency of the MRS is especially important in two scenarios"
("Specially" is incorrect here; "especially" is the right choice.)

"each of theses variations" → "each of these variations"
("theses" should be "these.")

"We begin by importing it along with other utilities" → "We begin by importing it along with other utilities."
(Missing period at the end.)

"An MultivariateRejectionSampler object is initialized" → "A MultivariateRejectionSampler object is initialized"
("An" should be "A" because "Multivariate" starts with a consonant sound.)

"Running a monte carlo simulation requires you to specifies" → "Running a Monte Carlo simulation requires you to specify"
("specifies" should be "specify.")

"values is a 2-tuple" → "value is a 2-tuple"
(Singular/plural agreement.)

"Let us take a look at some numbers. All tests ran in a Dell G15 5530, with 16 13th Gen Intel® Core™ i5-13450HX CPUs, 16Gb RAM, running ubuntu 22.04."
→ "Let us take a look at some numbers. All tests were run on a Dell G15 5530 with 16 13th Gen Intel® Core™ i5-13450HX CPUs, 16GB RAM, running Ubuntu 22.04."
("ran" → "were run," "16Gb" → "16GB," "ubuntu" → "Ubuntu.")

"then it might be better suited to re-run the simulations." → "then it might be better to re-run the simulations."
("better suited to" → "better to" for smoother phrasing.)
```

Gui-FernandesBR reviewed (approved):
LGTM.

Great implementation, @Lucas-Prates !

I would kindly recommend adding unit tests to some of the methods you've develop, just to make sure we cover some of the new code lines.

Gui-FernandesBR commented:
Also, please update the CHANGELOG file!

Lucas-Prates commented on docs/user/mrs.rst:
```diff
@@ -0,0 +1,230 @@
+===============================
+
+Multivariate Rejection Sampling allows you to quickly subsample the results of a 
+previous monte carlo simulation to obtain the results when one or more variables 
```
Yes, I forgot to solve that one! Thanks!

Lucas-Prates commented on rocketpy/simulation/multivariate_rejection_sampler.py:
```diff
@@ -0,0 +1,266 @@
+    inputs_json: dict = None
+    outputs_json: dict = None
+    probability_ratio: float = None
+    acceptance_probability: float = None
```
Python 3.9 does not support the use of "|" as short for "Union" in typing (https://docs.python.org/3.10/library/typing.html#typing.Union). Hence, I used "Union" for the moment and introduced a TODO once we no longer support python 3.9.

Lucas-Prates commented:
I have addressed the suggestions made by @Gui-FernandesBR, implemented unit and integration tests for the MRS, and updated the CHANGELOG.

Gui-FernandesBR reviewed (approved):
LGTM

Gui-FernandesBR commented:
@Lucas-Prates good work. Please squash and merge at your earliest convinience

MateusStano reviewed (commented):
This is really good overall! Great job

Gui-FernandesBR commented on rocketpy/plots/monte_carlo_plots.py:
```diff
@@ -195,3 +195,251 @@ def all(self, keys=None):
+            ax.add_artist(ell)
+
+        plt.scatter(0, 0, s=30, marker="*", color="black", label="Launch Point")
+        plt.scatter(
+            original_apogee_x,
+            original_apogee_y,
+            s=5,
+            marker="^",
+            color="green",
+            label="Original Apogee",
+        )
+        plt.scatter(
+            original_impact_x,
+            original_impact_y,
+            s=5,
+            marker="v",
+            color="blue",
+            label="Original Landing Point",
+        )
+
```
There's no such thing as "opposite colors", but matplotlib does offer a fair good guide to selecting colormaps: https://matplotlib.org/stable/users/explain/colors/colormaps.html

I suggest blue/orange or blue/red.

Ideally the user should be able to select their specific color... But I understand the current limitations.

MateusStano replied:
I guess the right term in complementary not "oposite". Anyway, I meant something like this:
![image](https://github.com/user-attachments/assets/8cc02f4a-1dfc-4bf5-a613-e156ef4357c6)

Lucas-Prates replied:
I will use then a "tetradic" color combination I found in this [site]( https://www.canva.com/colors/color-wheel/). Here is the pallete I got, just for reference:

![image](https://github.com/user-attachments/assets/96cc996c-a833-4412-a003-515a0501d9e3)

Lucas-Prates replied:
This is the tetradic colors for the ellipses plot:
![image](https://github.com/user-attachments/assets/e167185f-19c3-4609-ae46-01465bb06c86)

Gui-FernandesBR commented on docs/user/mrs.rst:
```diff
@@ -0,0 +1,230 @@
+class to possibly save time. It is highly recommended that you read about Monte 
+Carlo simulations.
+
+Motivation
+----------
+
+As discussed in :ref:`sensitivity-practical`, there are several sources of
+uncertainty that can affect the flight of a rocket, notably the weather and
+the measurements errors in design parameters. Still, it is desirable that the flight
+accomplishes its goal, for instance reaching a certain apogee, as well as staying under
+some safety restrictions, such as ensuring that the landing point is outside 
+of a given area.
+
+Monte Carlo simulation is a technique that allows us to quantify the uncertainty and
+give objective answers to those types of questions in terms of probabilities and 
+statistics. It relies on running several simulations under different conditions 
+specified by probability distributions provided by the user. Hence, depending on the
+inputs and number of samples, running these Monte Carlo simulations might take a while.
+
```
I'm afraid it is not possible to "align" or "justify" text in .rst files.

Of course it is always possible to include .html sections into the code, but I do not recommend that.

Honestly, I think we should let it as is, and let users complain whether the sections are too long.

Using "tips", "notes", and other special widgets usually helps the reader to navigate faster through the text, but it is totally understandable if we cannot provide them in some cases.

Lucas-Prates commented on docs/user/mrs.rst:
```diff
@@ -0,0 +1,230 @@
+second entry contains the pdf of the new distribution. The code below shows how to
+create these distributions and the dictionary
+
+.. jupyter-execute::
+
+    old_mass_pdf = norm(15.426, 0.5).pdf
+    new_mass_pdf = norm(15, 0.5).pdf
+    distribution_dict = {
+        "mass": (old_mass_pdf, new_mass_pdf),
+    }
+
```
The latest commit now makes this possible. As long as that fin set has a name different from the first one, say "Fins2", the key of the dictionary would be "aerodynamic_surfaces_Fins2_root_chord". The explanation is given in the mrs.rst file.

Lucas-Prates commented on rocketpy/plots/monte_carlo_plots.py:
```diff
@@ -195,3 +195,251 @@ def all(self, keys=None):
+        # if not isinstance(other_monte_carlo, MonteCarlo):
+        #     raise TypeError(
+        #         "Argument 'other_monte_carlo' must be an MonteCarlo object!"
+        #     )
```
This TODO refered to a condition that verified if the user was passing a `MonteCarlo` to the comparison. Perhaps it is not that important, as it might be intuitive and is written in the documentation. Moreover, I am having problems with the `isinstance(other_monte_carlo, MonteCarlo)` because, when the user uses the `import_results` method, the MonteCarlo object, for some reason, no longer satisfies the previous condition. I might be doing something terribly wrong, but I chose to remove the verification since it prohibited the comparison against a valid object.

Lucas-Prates commented:
This PR is ready again for review. Since last time:

1 - addressed the points in Stano's review about documentation and color;
2 - removed some input checks which I think were not that useful but were a bit trick to do correctly;
3 - modified the `flatten_dict` function to better handle variables in inner levels.

**The most important point is 3 since it might introduce a breaking change.**

Gui-FernandesBR commented on rocketpy/simulation/multivariate_rejection_sampler.py:
```diff
@@ -0,0 +1,292 @@
+
+
+@dataclass
+# TODO: replace Union by "|" once python 3.9 is no longer supported
```
What is the advantage of using `|` ? If none, I'd recommend removing the TODO.

```suggestion
```

An alternative is using the `Optional` type hint.

Gui-FernandesBR replied:
Btw, do you really need the `disable=unsupported-binary-operation` here?

Instead of defining it outside the class, why not defining it within the class context?

Lucas-Prates replied:
I do not think there is an advantage, the `|` operator is just a short hand for the `Union` introduced in [python 3.10](https://docs.python.org/3/library/typing.html#typing.Union).

Lucas-Prates replied:
The `disable=unsupported-binary-operation` was used when the `|` was used as well. In that case, pylint raises that flag, perhaps because of python 3.9. If we remove the TODO, we can remove both.

Gui-FernandesBR reviewed (approved):
All looking good to me, great implementation.
I don't see the changes to `flatten_dict` as a problem.
Nobody is using that function right now.

Waiting for @MateusStano 's final comments so we can proceed.

Gui-FernandesBR commented on rocketpy/plots/monte_carlo_plots.py:
```diff
@@ -195,3 +195,238 @@ def all(self, keys=None):
+        if image is not None:
+            try:
+                img = imageio.imread(image)
+            except FileNotFoundError as e:
```
```suggestion
            except FileNotFoundError as e:  # pragma no cover
```

Gui-FernandesBR commented on rocketpy/plots/monte_carlo_plots.py:
```diff
@@ -195,3 +195,238 @@ def all(self, keys=None):
+        if len(original_apogee_x) == 0 and len(original_impact_x) == 0:
+            raise ValueError("No apogee or impact data found. Cannot plot ellipses.")
+
+        original_impact_ellipses, original_apogee_ellipses = (
```
@copilot you are wrong, unfortunately
